codegenerator/cli/templates/static/codegen/src/db/EntityHistory.res

```diff
@@ -0,0 +1,325 @@
+  | Object({items}) =>
+    let nulldict = Js.Dict.empty()
+    items->Belt.Array.forEach(({location}) => {
+      nulldict->Js.Dict.set(location, %raw(`null`))
+    })
+    nulldict
+  | _ =>
+    Js.Exn.raiseError("Failed creating entityWithNullFields. Expected an object schema for entity")
+  }
+
+  //Gets an entity object with all fields set to null except for the id field
+  let getEntityWithNullFields = (entityId: string) => {
+    entityWithNullFields->Utils.Dict.updateImmutable("id", entityId->Utils.magic)
+  }
+
+  //Maps a schema object for the given entity with all fields nullable except for the id field
+  //Keeps any original nullable fields
+  let nullableEntitySchema: S.t<Js.Dict.t<unknown>> = S.schema(s =>
+    switch entitySchema->S.classify {
+    | Object({items}) =>
+      let nulldict = Js.Dict.empty()
+      items->Belt.Array.forEach(({location, schema}) => {
+        let nullableFieldSchema = switch (location, schema->S.classify) {
+        | ("id", _)
+        | (_, Null(_)) => schema //TODO double check this works for array types
+        | _ => S.null(schema)->S.toUnknown
+        }
+
+        nulldict->Js.Dict.set(location, s.matches(nullableFieldSchema))
+      })
+      nulldict
+    | _ =>
+      Js.Exn.raiseError(
+        "Failed creating nullableEntitySchema. Expected an object schema for entity",
+      )
+    }
+  )
+
+  let previousWithNullFields = {
+    chain_id: None,
+    block_timestamp: None,
+    block_number: None,
+    log_index: None,
+  }
+
+  S.object(s => {
+    {
+      "current": s.flatten(currentHistoryFieldsSchema),
+      "previous": s.flatten(previousHistoryFieldsSchema),
+      "entityData": s.flatten(nullableEntitySchema),
+      "action": s.field("action", Enums.EntityHistoryRowAction.enum.schema),
+    }
+  })->S.transform(s => {
+    parser: v => {
+      current: v["current"],
+      previous: switch v["previous"] {
+      | {
+          chain_id: Some(chain_id),
+          block_timestamp: Some(block_timestamp),
+          block_number: Some(block_number),
+          log_index: Some(log_index),
+        } =>
+        Some({
+          chain_id,
+          block_timestamp,
+          block_number,
+          log_index,
+        })
+      | {chain_id: None, block_timestamp: None, block_number: None, log_index: None} => None
+      | _ => s.fail("Unexpected mix of null and non-null values in previous history fields")
+      },
+      entityData: switch v["action"] {
+      | SET => v["entityData"]->S.parseAnyOrRaiseWith(entitySchema)->Set
+      | DELETE => v["entityData"]->S.parseAnyOrRaiseWith(entityIdOnlySchema)->Delete
+      },
+    },
+    serializer: v => {
+      let (entityData, action) = switch v.entityData {
+      | Set(entityData) => (
+          entityData->(Utils.magic: 'entity => Js.Dict.t<unknown>),
+          Enums.EntityHistoryRowAction.SET,
+        )
+      | Delete({id}) => (getEntityWithNullFields(id), DELETE)
```

Comment:
I don't really like that it'll create an array to return the tuple from the switch, but this is not super critical

Reply:
I can inline it and just have 2 switch statements. Not sure if array allocation or 2 switches is worse

Reply:
Just a quick benchmark for this and they look equivalent to me: 

[rescript code](https://rescript-lang.org/try?version=v11.1.4&code=C4TwDgpgBAhgxsAlgewHZQLxQIJQD5QBCAUMaJFAG4wA2mUAkqsABQDeqArgLYBcUiZgF8AlPigBlYACd2AZxn8F0wQHNRZcNEHAI01LVIQAHrv20o3GKsRx+AchiYAfFHsAjegCIApIgAmEMyIoF6k5NDSyADu9GzEULAIKKj88EhoADQJVLT8OnoGNNlCpDQQwFDI7gBWACrIAEoxuFgs1DT8HSL8UbEYrvGJ5ZUs6SmZuTRiWHLRIXAAFlNQQ+JMrBw8+aIuUCzYk4gAtM5WNnAiOQRSsmzKSrsD+4STcqfntleJpTlrieMsjlEh1sj9iL8RlVag1mtFCPR2nkpj0oH09mtAakoHMFssOqtroxmCwAPozVzYIm3MkUog5IRgqZKebAJYrNYEDbsLh8ARPVwnM7WWzUmTyRQ4gU4j4iuAMkplCrQmobejcrZ8ACMGih1Rqt3oNPukq8ADNkMgvBogA)

[benchmark](https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJWdEx6SzlYNG9fejF0WXQwSG43VWIiLCJjb2RlIjoiREFUQS5vYmpUb1Jvd0EoREFUQS5vYmpJbnQpIiwiZGVwZW5kZW5jaWVzIjpbXX0seyJpZCI6Ii1xTDJleHRvYlBMM2xHSUw2NVZCdiIsImNvZGUiOiJEQVRBLm9ialRvUm93QihEQVRBLm9iakludCkiLCJkZXBlbmRlbmNpZXMiOltdfSx7ImlkIjoiNF9taGZ4SkVMZUpNR0ZvQnc1dU5qIiwiY29kZSI6IkRBVEEub2JqVG9Sb3dBKERBVEEub2JqU3RyKSIsImRlcGVuZGVuY2llcyI6W119LHsiaWQiOiJXRGdvS1ItMmRydkVNV3dycmtEajkiLCJjb2RlIjoiREFUQS5vYmpUb1Jvd0IoREFUQS5vYmpTdHIpIiwiZGVwZW5kZW5jaWVzIjpbXX1dLCJjb25maWciOnsibmFtZSI6IkJhc2ljIGV4YW1wbGUiLCJwYXJhbGxlbCI6dHJ1ZSwiZ2xvYmFsVGVzdENvbmZpZyI6eyJkZXBlbmRlbmNpZXMiOltdfSwiZGF0YUNvZGUiOiJmdW5jdGlvbiBvYmpUb1Jvd0EodmFsKSB7XG4gIHZhciBtYXRjaDtcbiAgbWF0Y2ggPSB2YWwuVEFHID09PSBcIkludFwiID8gW1xuICAgICAgXCJBXCIsXG4gICAgICB2YWwubnVtXG4gICAgXSA6IFtcbiAgICAgIFwiQlwiLFxuICAgICAgdmFsLnN0clxuICAgIF07XG4gIHJldHVybiB7XG4gICAgICAgICAgYWN0aW9uOiBtYXRjaFswXSxcbiAgICAgICAgICB2YWw6IG1hdGNoWzFdXG4gICAgICAgIH07XG59XG5cbmZ1bmN0aW9uIG9ialRvUm93Qih2YWwpIHtcbiAgdmFyIHRtcDtcbiAgdG1wID0gdmFsLlRBRyA9PT0gXCJJbnRcIiA_IFwiQVwiIDogXCJCXCI7XG4gIHZhciB0bXAkMTtcbiAgdG1wJDEgPSB2YWwuVEFHID09PSBcIkludFwiID8gdmFsLm51bSA6IHZhbC5zdHI7XG4gIHJldHVybiB7XG4gICAgICAgICAgYWN0aW9uOiB0bXAsXG4gICAgICAgICAgdmFsOiB0bXAkMVxuICAgICAgICB9O1xufVxuXG52YXIgb2JqSW50ID0ge1xuICBUQUc6IFwiSW50XCIsXG4gIG51bTogMVxufTtcblxudmFyIG9ialN0ciA9IHtcbiAgVEFHOiBcIlN0clwiLFxuICBzdHI6IFwiZm9vXCJcbn07XG5cbnJldHVybiB7XG4gIG9ialRvUm93QSAsXG4gIG9ialRvUm93QiAsXG4gIG9iakludCAsXG4gIG9ialN0ciAsXG59In19)

Reply:
Ok, nice, I won't care about cases like this anymore